src/Http/Routing/src/Internal/LinkGenerationDecisionTree.cs

```diff
@@ -91,7 +108,20 @@ public IList<OutboundMatchResult> GetMatches(RouteValueDictionary values, RouteV
                 {
                     if (criterion.Branches.TryGetValue(value ?? string.Empty, out var branch))
                     {
-                        Walk(results, values, ambientValues, branch, isFallbackPath);
+                        Walk(results, values, ambientValues, branch, quality + 1m);
+                    }
+                    else
+                    {
+                        // If an explicitly specified value doesn't match any branch, then speculatively walk the
+                        // "null" path if the value doesn't match any known value.
+                        //
+                        // This can happen when linking from a page <-> action. We want to be
+                        // able to use "page" and "action" as normal route parameters.
+                        var knownValues = _knownValues[key];
+                        if (!knownValues.Contains(value ?? string.Empty) && criterion.Branches.TryGetValue(string.Empty, out branch))
+                        {
+                            Walk(results, values, ambientValues, branch, quality + 0.01m);
+                        }
```

Comment:
This is the gist of the fix. Each approach has problems. This kind of feature as at-odds with the algorithmic requirements of routing so it's going to require esoteric heuristics to improve this.

=====

A few things we could try:

1. (in this pr) check 'page' against known values to prevent ambiguity
1. allow the tree to ignore 'page' and rely on ordering - changes the behavior when you have a link fail to generate
1. remove 'page' from required values of an action

I went with approach once because I don't like the drawbacks of 2. I don't want to impact the majority cases by changing how it behaves when you fail to generate a link when linking from a page to an action.

Imagine you're trying to like an action: `{ action = "Buy", controller = "Store" }` with template `/Store/Buy/{productId?}`. you are on page `/Foo/Bar` and you supply the route values `{ action = "Buy", controller = "Store" }` - but no productId. If we did option 2, then both your action *and your current page* would be candidates. Since you failed to supply productId, the action would fail to generate, and you'd get `/Foo/Bar?action=Buy&controller=Store` instead.

This is why I did option 1. 

Option 3 is totally busto because it makes ambient value invalidation stop working when you like from a page to an action. Yuck.

Reply:
Makes sense. I think I tried option 3 before realizing I was out of my depth 😄